src/rust/ide/src/controller/project.rs

```diff
@@ -304,6 +312,20 @@ impl Handle {
     pub fn register_execution_context(&self, execution_context:&Rc<ExecutionContext>) {
         self.execution_contexts.insert(execution_context.clone_ref());
     }
+
+    /// Get project's name.
+    pub fn project_name(&self) -> ProjectName {
+        self.project_metadata.borrow().name.clone()
+    }
+
+    /// Rename project.
+    pub async fn rename_project(&self, name:impl Str) -> FallibleResult<()> {
+        let name                 = name.into();
+        let mut project_metadata = self.project_metadata.borrow_mut();
+        self.project_manager.rename_project(&project_metadata.id,&name).await?;
+        *project_metadata.name = ProjectName::new(name).to_string();
```

Comment:
This looks *very scary*. `.await` yields the control while we have ongoing `borrow_mut`.
If we request a rename, then try reading the name while waiting for engine's answer, this will likely panic.

We must always try to keep refcell's borrows as localized as possible and avoid allowing non-local code execution while we have lock in place.

Reply:
Yep, we should never have such code. Borrow must should be always limited to a narrow scope.

Reply:
Good catch! Thanks for pointing it out.